### PR TITLE
Publish and Subscribe via Kafka

### PR DESCRIPTION
**Pull request by michaeldauria**
We use a ConsumerGroup on the subscriber side to allow for multiple consumers per topic.

### CODE REVIEW

File: lib/promiscuous/cli.rb [hunk 1]
```diff
@@ -71,7 +71,7 @@ def record
   def replay_payload(payload)
     endpoint = MultiJson.load(payload)['__amqp__']
     if endpoint
-      # TODO confirm
+      Promiscuous::Kafka.publish(:key => endpoint, :payload => payload)
```
Comment (michaeldauria) [inline lib/promiscuous/cli.rb]: this `:key` is incorrect

File: lib/promiscuous/config.rb [hunk 2]
```diff
@@ -1,21 +1,26 @@
...
+                 :socket_timeout, :heartbeat, :sync_all_routing, :prefetch,
+                 :publisher_lock_expiration, :publisher_lock_timeout,
+                 :recovery_interval, :logger, :subscriber_threads, :version_field,
+                 :error_notifier, :test_mode, :on_stats, :max_retries, :generation,
+                 :destroy_timeout, :destroy_check_interval,
+    # vvvv REMOVE AFTER AMQP IS GONE vvvv
+                 :backend, :amqp_url, :publisher_amqp_url, :subscriber_amqp_url,
+                 :publisher_exchange, :subscriber_exchanges, :sync_exchange, :queue_name,
+                 :queue_options, :error_queue_name, :rabbit_mgmt_url, :queue_policy,
                  :error_exchange, :error_routing, :retry_routing, :error_ttl
```
Comment (michaeldauria) [inline lib/promiscuous/config.rb]: `:error_ttl` will live on

File: lib/promiscuous/config.rb [hunk 3]
```diff
@@ -1,21 +1,26 @@
...
 
   def self.backend=(value)
     @@backend = value
     Promiscuous::AMQP.backend = value
+    Promiscuous::Kafka.backend = (value == :bunny)? :poseidon : value
```
Comment (michaeldauria) [inline lib/promiscuous/config.rb]: TODO: refactor to allow for either `bunny` or `kafka` for subscription and both for publishing

File: lib/promiscuous/config.rb [hunk 4]
```diff
@@ -63,8 +77,8 @@ def self._configure(&block)
     self.publisher_lock_expiration ||= 5.seconds
     self.publisher_lock_timeout ||= 2.seconds
     self.recovery_interval    ||= 5.seconds
-    self.logger               ||= Logger.new(STDERR).tap { |l| l.level = ENV["LOGGER_LEVEL"].to_i || Logger::WARN }
-    self.subscriber_threads   ||= 10
+    self.logger               ||= defined?(Rails) ? Rails.logger : Logger.new(STDERR).tap { |l| l.level = Logger::WARN }
+    self.subscriber_threads   ||= 1
```
Comment (michaeldauria) [inline lib/promiscuous/config.rb:79]: bump back up to 10?

File: lib/promiscuous/kafka/poseidon.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
+require 'poseidon'
+require 'poseidon_cluster'
+
+class Promiscuous::Kafka::Poseidon
+
+  # TODO: Move to spec support
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: move to spec/support

File: lib/promiscuous/kafka/poseidon.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    # TODO: client id needs to be unique across the system, use hostname+rand?
+    client_id = ['promiscuous', Promiscuous::Config.app, Poseidon::Cluster.guid].join('.')
+    connection = ::Poseidon::Producer.new(Promiscuous::Config.kafka_hosts, client_id,
+                                          :type => :sync,
+                                          :compression_codec => :none,
+                                          :metadata_refresh_interval_ms => 600_000,
+                                          :max_send_retries => 10,
+                                          :retry_backoff_ms => 100,
+                                          :required_acks => 1,
+                                          :ack_timeout_ms => 1000,
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: most, if not all, of these need to be configurable

File: lib/promiscuous/kafka/poseidon.rb [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                                          :socket_timeout_ms => Promiscuous::Config.socket_timeout)
+    connection
+  end
+
+  def connect
+    @connection = new_connection
+    # Poseidon.logger = Logger.new(STDOUT).tap { |l| l.level = 0 }
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: move to spec/support

File: lib/promiscuous/kafka/poseidon.rb [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+  def disconnect
+    @connection_lock.synchronize do
+      return unless connected?
+      @connection.shutdown
+      @connection = nil
+    end
+  end
+
+  #TODO: broken
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: broken

File: lib/promiscuous/kafka/poseidon.rb [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  def connected?
+    !!@connection.try(:connected?)
+  end
+
+  def raw_publish(options)
+    ok = false
+    10.times do
+      ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+      break if ok
+      sleep(1)
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: this might be too long to `sleep`

File: lib/promiscuous/kafka/poseidon.rb [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      @consumer.fetch(:commit => false) do |partition, payloads|
+        Promiscuous.debug "[kafka] Received #{payloads.count} payloads topic:#{@consumer.topic} #{Thread.current}"
+        payloads.each do |payload|
+          Promiscuous.debug "[kafka] Fetched '#{payload.value}' topic:#{@consumer.topic} offset:#{payload.offset} parition:#{partition}"
+          block.call(MetaData.new(@consumer, partition, payload.offset), payload)
+        end
+      end
+    end
+
+    def delete_queues
```
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: remove this method, we fast forward the topic instead

File: lib/promiscuous/publisher/operation/base.rb [hunk 11]
```diff
@@ -101,15 +101,15 @@ def queue_operation_payloads(operations = self.operations)
...
+    @operation_payloads.map do |lock_key, operation_payloads|
       payload              = {}
       payload[:operations] = operation_payloads
       payload[:app]        = Promiscuous::Config.app
       payload[:timestamp]  = Time.now
       payload[:generation] = Promiscuous::Config.generation
       payload[:host]       = Socket.gethostname
+      payload[:key]        = lock_key
       payload.merge!(self.payload_attributes)
-      MultiJson.dump(payload)
```
Comment (michaeldauria) [inline lib/promiscuous/publisher/operation/base.rb:112]: if there is a way to get at this `lock_key` from the `instance` we can put this back and refactor

File: lib/promiscuous/subscriber/unit_of_work.rb [hunk 12]
```diff
@@ -40,8 +40,9 @@ def process_message
       on_message
     rescue Exception => e
       Promiscuous::Config.error_notifier.call(e)
-      message.nack
+      # message.nack
       raise e if Promiscuous::Config.test_mode
+      sleep Promiscuous::Config.error_ttl / 1000.0; retry
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/unit_of_work.rb]: use the `sleep` for Kafka and the `nack` for Rabbit

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  class DistributorThread
+    def initialize(topic)
+      # late include of CelluloidSubscriber because the class is resolved
+      # at runtime since we can have different backends.
+      extend Promiscuous::Kafka::Subscriber
+
+      @stop = false
+      @topic = topic
+      @thread = Thread.new { main_loop }
+      # @thread.abort_on_exception = true
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/distributor.rb]: remove

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+      Promiscuous.warn "[kafka] [receive] cannot process message: #{e}\n#{e.backtrace.join("\n")}"
+      Promiscuous::Config.error_notifier.call(e)
+    end
+
+    # TODO: make sure we're on the sync topic(s) as well
+    def main_loop
+      @consumer = subscribe(@topic)
+      while not @stop do
+        begin
+          fetch_and_process_messages(&method(:on_message))
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/distributor.rb:52]: this should read `@consumer.fetch_and_process_messages` for clarity

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        rescue Poseidon::Connection::ConnectionFailedError
+          @consumer = subscribe(@topic)
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/distributor.rb:55]: will this cause a rebalance and leave the replaced consumer in a broken state?

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    def stop
+      Promiscuous.debug "[distributor] stopping status:#{@thread.status} #{@thread}"
+
+      # We wait in case the consumer is responsible for more than one partition
+      # see: https://github.com/bsm/poseidon_cluster/blob/master/lib/poseidon/consumer_group.rb#L229
+      @stop = true
+      @thread.join
+    end
+
+    def show_stop_status(num_requests)
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/distributor.rb:73]: not sure where `num_requests` comes from

File: lib/promiscuous/subscriber/worker/recorder.rb [hunk 17]
```diff
@@ -1,20 +1,13 @@
 class Promiscuous::Subscriber::Worker::Recorder
   def initialize(log_file)
     @log_file = log_file
-    extend Promiscuous::AMQP::Subscriber
+    extend Promiscuous::Kafka::Subscriber
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/recorder.rb]: we need to support both backends here

File: spec/integration/consistency_spec.rb [hunk 18]
```diff
@@ -46,17 +46,17 @@
       pub = PublisherModel.create(:field_1 => '1')
       eventually { SubscriberModel.count.should == 1 }
 
-      amqp_delayed!
+      kafka_delayed!
```
Comment (michaeldauria) [inline spec/integration/consistency_spec.rb]: `backend_*`, since they function identically

File: spec/integration/payload_spec.rb [hunk 19]
```diff
@@ -10,7 +10,7 @@
 
     PublisherModel.create(:field_1 => '1')
 
-    Promiscuous::AMQP::Fake.get_next_payload['host'].should == 'example.com'
+    Promiscuous::Kafka::Fake.get_next_payload['host'].should == 'example.com'
```
Comment (michaeldauria) [inline spec/integration/payload_spec.rb]: `Promiscuous::Backend::Fake`

File: spec/integration/retry_exceptions_spec.rb [hunk 20]
```diff
@@ -38,7 +38,7 @@
     end
 
     context 'when the ttl is longer then the subscriber is expected to receive the message' do
-      let(:ttl) { 99999 }
+      let(:ttl) { 2000 }
```
Comment (michaeldauria) [inline spec/integration/retry_exceptions_spec.rb:42]: this number is actually in ms and not seconds

File: spec/support/backend.rb [hunk 21]
```diff
@@ -27,6 +29,7 @@ def use_real_backend(options={}, &block)
         block.call(config) if block
       end
     end
+    Promiscuous::Kafka::Poseidon.advance_offsets_forward!
```
Comment (michaeldauria) [inline spec/support/backend.rb]: move this into support somewhere

File: promiscuous.gemspec [hunk 22]
```diff
@@ -20,6 +20,8 @@ Gem::Specification.new do |s|
   s.add_dependency "activemodel",         ">= 3"
   s.add_dependency "bunny",               ">= 0.10.7"
   s.add_dependency "amq-protocol",        ">= 1.8.0"
+  s.add_dependency "poseidon",            "~> 0.0.5"
+  s.add_dependency "promiscuous-poseidon_cluster",    "~> 0.3.0 "
   s.add_dependency "ruby-progressbar",    "~> 1.2.0"
   s.add_dependency "redis",               "~> 3.0"
   s.add_dependency "algorithms",          "~> 0.6.1"
```
Comment (michaeldauria) [inline promiscuous.gemspec]: do we need this `algorithms` gem anymore?

File: lib/promiscuous.rb [hunk 23]
```diff
@@ -44,6 +44,7 @@ def configure(&block)
 
     def connect
       AMQP.connect
+      Kafka.connect # TODO: needed?
```
Comment (kareemk) [inline lib/promiscuous.rb]: Figure this out.

File: lib/promiscuous.rb [hunk 24]
```diff
@@ -51,6 +52,7 @@ def connect
 
     def disconnect
       AMQP.disconnect
+      Kafka.disconnect
```
Comment (kareemk) [inline lib/promiscuous.rb]: Also is this needed?

File: lib/promiscuous/config.rb (hunk 4 above)
Comment (kareemk) [inline lib/promiscuous/config.rb:79]: Yes. And we should run tests with 10 to make sure everything still works.

File: lib/promiscuous/kafka/file.rb [hunk 25]
```diff
@@ -0,0 +1,78 @@
+class Promiscuous::Kafka::File
```
Comment (kareemk) [inline lib/promiscuous/kafka/file.rb]: file?

File: lib/promiscuous/kafka/poseidon.rb [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+    broker_pool.update_known_brokers({ 0 => { :host => broker_host, :port => broker_port }})
+
+    # we assume that a topic maps to a ConsumerGroup one-to-one
+    zk = ZK.new(Promiscuous::Config.zookeeper_hosts.join(','))
+    begin
+      Promiscuous::Config.subscriber_topics.each do |topic|
+        partitions_path = "/consumers/#{topic}/offsets/#{topic}"
+        zk.children(partitions_path).each do |partition|
+          partition_offset_requests = [::Poseidon::Protocol::PartitionOffsetRequest.new(partition.to_i, -1, 1000)]
+          offset_topic_requests = [::Poseidon::Protocol::TopicOffsetRequest.new(topic, partition_offset_requests)]
```
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: Clean this all up somehow. It's very hard to follow.

File: lib/promiscuous/kafka/poseidon.rb [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+
+  attr_accessor :connection, :connection_lock
+
+  def initialize
+    # The poseidon socket doesn't like when multiple threads access to it apparently
+    @connection_lock = Mutex.new
+  end
+
+  def new_connection
+    # TODO: client id needs to be unique across the system, use hostname+rand?
```
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: seems important

File: lib/promiscuous/kafka/poseidon.rb (hunk 6 above)
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: yes, and really understand the implications of each.

File: lib/promiscuous/kafka/poseidon.rb [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+  end
+
+  #TODO: broken
+  def connected?
+    !!@connection.try(:connected?)
+  end
+
+  def raw_publish(options)
+    ok = false
+    10.times do
```
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: why?

File: lib/promiscuous/kafka/poseidon.rb [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+    def subscribe(topic)
+      @consumer = ::Poseidon::ConsumerGroup.new(Promiscuous::Config.app,
+                                                Promiscuous::Config.kafka_hosts,
+                                                Promiscuous::Config.zookeeper_hosts,
+                                                topic, :trail => Promiscuous::Config.test_mode, :max_wait_ms => 10)
+    end
+
+    def fetch_and_process_messages(&block)
+      Promiscuous.debug "[kafka] Fetching messages topic:#{@consumer.topic} #{Thread.current}"
+      @consumer.fetch(:commit => false) do |partition, payloads|
```
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: Overall you need to space out your lines more. Makes it much easier to read.

File: lib/promiscuous/kafka/poseidon.rb [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+      def initialize(consumer, partition, offset)
+        @consumer = consumer
+        @partition = partition
+        @offset = offset
+
+        Promiscuous.debug "[kafka] [metadata] topic:#{@consumer.topic} offset:#{offset} partition:#{partition}"
+      end
+
+      # TODO: rename to advance_offset or commit?
+      def ack
```
Comment (kareemk) [inline lib/promiscuous/kafka/poseidon.rb]: probably ok as `ack` for now

File: lib/promiscuous/publisher/operation/base.rb (hunk 11 above)
Comment (kareemk) [inline lib/promiscuous/publisher/operation/base.rb:112]: there's definitely a way to refactor this. i'll think about it.

File: lib/promiscuous/kafka/file.rb (hunk 25 above)
Comment (michaeldauria) [inline lib/promiscuous/kafka/file.rb]: this is for the file backend which is what powers the recorder. Best bet here is to have a generic Backend class instead of one per driver which is what it is now

File: lib/promiscuous/kafka/poseidon.rb (hunk 27 above)
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: yea, i had taken care of this and forgot to remove the comment

File: lib/promiscuous/kafka/poseidon.rb (hunk 28 above)
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: we had taken this from that test suite to make sure that messages were being published. We should probably try to take it out and see how it behaves.

File: lib/promiscuous/kafka/poseidon.rb (hunk 29 above)
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: not sure what you mean, example? I thought we were supposed to line up all of the arguments with the first, which is what I had done here.

File: lib/promiscuous/kafka/poseidon.rb (hunk 30 above)
Comment (michaeldauria) [inline lib/promiscuous/kafka/poseidon.rb]: It has to remain this way if we're going to have pluggable backends with the same interface

File: lib/promiscuous.rb (hunk 24 above)
Comment (michaeldauria) [inline lib/promiscuous.rb]: this is yes, so we cleanly shutdown the connection to Kafka and ZK

File: .travis.yml [hunk 31]
```diff
@@ -1,11 +1,13 @@
 language: ruby
 bundler_args: --without development
 script: bundle exec rspec spec
-env: TIMEOUT=10
+env:
+  - TIMEOUT=10
+  - PROMISCUOUS_BACKEND=bunny
```
Comment (michaeldauria) [inline .travis.yml]: this will become poseidon once i can get my own Kafka cluster started and the tests running on it

File: lib/promiscuous/subscriber/unit_of_work.rb [hunk 32]
```diff
@@ -36,12 +36,25 @@ def self.current=(value)
...
+    retry_max = 50
+
     begin
       on_message
     rescue Exception => e
       Promiscuous::Config.error_notifier.call(e)
-      message.nack
       raise e if Promiscuous::Config.test_mode
+
+      # TODO: use dependency injection here instead
```
Comment (michaeldauria) [inline lib/promiscuous/subscriber/unit_of_work.rb]: or some other cleaner approach

File: lib/promiscuous/backend/poseidon.rb [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+    @connection = ::Poseidon::Producer.new(Promiscuous::Config.kafka_hosts, client_id,
+                                          :type => :sync,
+                                          :compression_codec => :none,
+                                          :metadata_refresh_interval_ms => 600_000,
+                                          :max_send_retries => 10,
+                                          :retry_backoff_ms => 100,
+                                          :required_acks => 1,
+                                          :ack_timeout_ms => 1000,
+                                          :socket_timeout_ms => Promiscuous::Config.socket_timeout)
+    @connection
```
Comment (benseligman) [inline lib/promiscuous/backend/poseidon.rb]: not necessary -- the assignment will return the value

File: lib/promiscuous/backend/poseidon.rb [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+    @connection_lock.synchronize do
+      return unless connected?
+      @connection.shutdown
+      @connection = nil
+    end
+  end
+
+  # TODO: extend Poseidon with a connected? method
+  def connected?
+    !@connection.nil?
```
Comment (benseligman) [inline lib/promiscuous/backend/poseidon.rb]: `@connection.present?`

File: lib/promiscuous/backend.rb [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+  autoload :Bunny, :Poseidon, :Null, :Fake, :File
+
+  class << self
+    attr_accessor :driver
+    attr_accessor :driver_class
+    attr_accessor :subscriber_worker_module
+
+    def driver=(value)
+      disconnect
+      @driver_class = value.nil? ? nil : "Promiscuous::Backend::#{value.to_s.camelize.gsub(/backend/, 'Backend')}".constantize
```
Comment (benseligman) [inline lib/promiscuous/backend.rb]: not a big deal, but i find `value.try { |v| "Promiscuous::Backend::#{value.to_s.camelize.gsub(/backend/, 'Backend')}".constantize }` a little easier to read

File: lib/promiscuous/backend/poseidon.rb [hunk 36]
```diff
@@ -0,0 +1,128 @@
...
+      end
+
+      # TODO: rename to advance_offset or commit?
+      def ack
+        Promiscuous.debug "[kafka] [commit] topic:#{@consumer.topic} offset:#{@offset+1} partition:#{@partition}"
+        @consumer.commit(@partition, @offset+1)
+      end
+    end
+
+    module Worker
```
Comment (benseligman) [inline lib/promiscuous/backend/poseidon.rb]: just noticing that the worker just delegates to the distributor -- is there a need for both?

File: lib/promiscuous/subscriber/unit_of_work.rb (hunk 32 above)
Comment (benseligman) [inline lib/promiscuous/subscriber/unit_of_work.rb]: i like the approach we were talking about of moving `process_message` into each backend

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 37]
```diff
@@ -0,0 +1,88 @@
+class Promiscuous::Subscriber::Worker::Distributor
+  def initialize(root)
+    @root = root
+  end
+
+  def start
+    num_threads = Promiscuous::Config.subscriber_threads
+    Promiscuous::Config.subscriber_topics.each do |topic|
+      @distributor_threads ||= num_threads.times.map { DistributorThread.new(topic) }
```
Comment (benseligman) [inline lib/promiscuous/subscriber/worker/distributor.rb]: why is this cached?

File: lib/promiscuous/subscriber/worker/distributor.rb [hunk 38]
```diff
@@ -0,0 +1,88 @@
...
+    @distributor_threads = nil
+  end
+
+  def show_stop_status(num_requests)
+    @distributor_threads.to_a.each { |distributor_thread| distributor_thread.show_stop_status(num_requests) }
+  end
+
+  class DistributorThread
+    def initialize(topic)
+      # late include of CelluloidSubscriber because the class is resolved
```
Comment (benseligman) [inline lib/promiscuous/subscriber/worker/distributor.rb:30]: haha thanks for the comment -- it took me and @TomHoss a while to understand why another extend in promiscuous was inside initialize

File: promiscuous.gemspec (hunk 22 above)
Comment (benseligman) [inline promiscuous.gemspec]: did you try removing it?

File: spec/integration/amqp_spec.rb [hunk 39]
```diff
@@ -1,11 +1,13 @@
 require 'spec_helper'
 
-describe Promiscuous::AMQP do
-  before { use_real_backend }
+if Promiscuous::Config.backend == :bunny
```
Comment (benseligman) [inline spec/integration/amqp_spec.rb:3]: why this conditional in tests?

Comment (benseligman) [inline spec/integration/amqp_spec.rb:3]: nvm i understand

File: lib/promiscuous/backend.rb (hunk 35 above)
Comment (michaeldauria) [inline lib/promiscuous/backend.rb]: agreed

File: lib/promiscuous/subscriber/worker/distributor.rb (hunk 37 above)
Comment (michaeldauria) [inline lib/promiscuous/subscriber/worker/distributor.rb]: In case you call start twice by mistake, i guess that's defensive coding but i took the idea from the runner

File: .travis.yml [hunk 40]
```diff
@@ -1,11 +1,13 @@
 language: ruby
 bundler_args: --without development
 script: bundle exec rspec spec
-env: TIMEOUT=10
+env:
+  - TIMEOUT=10 PROMISCUOUS_BACKEND=bunny
+  - TIMEOUT=10 PROMISCUOUS_BACKEND=poseidon
```
Comment (kareemk) [inline .travis.yml]: BACKEND instead of PROMISCUOUS_BACKEND

File: lib/promiscuous/backend.rb (hunk 35 above)
Comment (kareemk) [inline lib/promiscuous/backend.rb]: you shouldn't have to specify namespace as its in the same namespace. also that gsub is dirty. remove backend completely.

File: lib/promiscuous/backend/poseidon.rb [hunk 41]
```diff
@@ -0,0 +1,128 @@
...
+  end
+
+  # TODO: extend Poseidon with a connected? method
+  def connected?
+    !@connection.nil?
+  end
+
+  def raw_publish(options)
+    ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+    raise "Unable to send messages" if !ok
```
Comment (kareemk) [inline lib/promiscuous/backend/poseidon.rb]: needs to be a Promiscuous::Error subclass

File: lib/promiscuous/backend/poseidon.rb [hunk 42]
```diff
@@ -0,0 +1,128 @@
...
+    Promiscuous.debug "[publish] [kafka] #{options[:topic]}/#{options[:topic_key]} #{options[:payload]}"
+  end
+
+  def publish(options={})
+    @connection_lock.synchronize do
+      raw_publish(options)
+      # TODO: Rabbit is in control of the unlocking for now
```
Comment (kareemk) [inline lib/promiscuous/backend/poseidon.rb]: how is this working then in tests?

File: lib/promiscuous/subscriber/unit_of_work.rb (hunk 32 above)
Comment (kareemk) [inline lib/promiscuous/subscriber/unit_of_work.rb]: this has to move out. yes, a process_message in backend makes sense.

File: lib/promiscuous/subscriber/worker.rb [hunk 43]
```diff
@@ -1,33 +1,37 @@
...
-  autoload :Pump, :Runner, :Stats, :Recorder, :EventualDestroyer
+  autoload :Pump, :Distributor, :Runner, :Stats, :Recorder, :EventualDestroyer
 
   attr_accessor :pump, :runner, :stats, :eventual_destroyer
 
   def initialize
-    @pump = Pump.new(self)
-    @runner = Runner.new(self)
+    # inject what we need for our backend
+    extend Promiscuous::Backend.subscriber_worker_module
```
Comment (kareemk) [inline lib/promiscuous/subscriber/worker.rb]: subscriber_methods

File: spec/support/backend.rb [hunk 44]
```diff
@@ -1,5 +1,41 @@
+require 'securerandom'
+
 module BackendHelper
+  def advance_offsets_forward!
```
Comment (kareemk) [inline spec/support/backend.rb]: this should be in KafkaHelper

File: spec/support/backend.rb [hunk 45]
```diff
@@ -13,24 +49,37 @@ def reconfigure_backend(&block)
       config.destroy_check_interval = 0
       config.max_retries = 0
       config.rabbit_mgmt_url = rabbit_mgmt_url
+      # config.kafka_hosts = kafka_hosts
```
Comment (kareemk) [inline spec/support/backend.rb]: why commented out?

File: spec/support/backend.rb [hunk 46]
```diff
@@ -73,14 +122,34 @@ def amqp_url
   def rabbit_mgmt_url
     ENV['BOXEN_RABBITMQ_MGMT_URL'] || 'http://guest:guest@localhost:15672'
   end
+
+  def kafka_hosts
```
Comment (kareemk) [inline spec/support/backend.rb]: you should split out the helpers for rabbit vs. kafka.

File: lib/promiscuous/backend/poseidon.rb (hunk 42 above)
Comment (michaeldauria) [inline lib/promiscuous/backend/poseidon.rb]: This was a bad comment and will be addressed with the `:both` backend i am working on

**Comment by benseligman** [comment]
lgtm
